model/ftn/wmgridmd.ftn

```diff
@@ -4721,6 +4759,684 @@
+!     !/SHRD Shared memory case.
+!
+! 10. Source code :
+!
+!/ ------------------------------------------------------------------- /
+!
+      USE CONSTANTS
+      USE W3GDATMD
+      USE W3ODATMD
+      USE W3ADATMD
+      USE WMMDATMD
+!
+      USE W3SERVMD, ONLY: EXTCDE
+!/SMC      USE W3PSMCMD, ONLY: W3SMCGMP, W3SMCELL
+
+!/S      USE W3SERVMD, ONLY: STRACE
+!
+      IMPLICIT NONE
+!
+!/MPI      INCLUDE "mpif.h"
+!/
+!/ ------------------------------------------------------------------- /
+!/ Parameter list
+!/
+!/ ------------------------------------------------------------------- /
+!/ Local parameters
+!/
+      INTEGER                :: I, J, IX, IY, IXY, JX, JY, NPJ,      &
+                                NR, NT, NA, NTL, JJ, NIT, NG, NOUT,  &
+                                ISEA, JSEA, IPRC, ITAG, TGRP, NPMX,  &
+                                IP, NP, ICROOT, JCROOT, IEER
+
+      INTEGER, Dimension(MPI_STATUS_SIZE):: MPIState
+
+!/S      INTEGER, SAVE            :: IENT = 0
+      INTEGER, ALLOCATABLE   :: NREC(:),  NSND(:),  NTPP(:),  &
+                                IBPTS(:), JBPTS(:), IPBPT(:)
+      REAL, PARAMETER        :: ODIMAX = 25.
+      REAL, ALLOCATABLE      :: XLon(:), YLat(:)
+      LOGICAL                :: CHANGE 
+      LOGICAL, ALLOCATABLE   :: SHRANK(:,:), DOGRID(:) 
+!/T5      CHARACTER(LEN=18), ALLOCATABLE :: TSTR(:)
+!/T5      CHARACTER(LEN=18)              :: DSTR
+!
+      TYPE STORE
+        INTEGER              :: NTOT, NFIN
+        INTEGER, POINTER     :: ICVBP(:), MSDBP(:), ISS(:), JSS(:),   &
+                                JCVBP(:), IPCVB(:), IPS(:), ITG(:)
+        LOGICAL, POINTER     :: FLA(:)
+        LOGICAL              :: INIT
+      END TYPE STORE
+!
+      TYPE(STORE), ALLOCATABLE :: STORES(:,:)
+!/
+!/S      CALL STRACE (IENT, 'WMSMCEQL ')
+!
+! -------------------------------------------------------------------- /
+! 0.  Initializations
+!
+      ALLOCATE ( SHRANK(NRGRD,NRGRD), STORES(NRGRD,NRGRD),            &
+                 DOGRID(NRGRD), STAT=ISTAT )
+      CHECK_ALLOC_STATUS ( ISTAT )
+!
+      SHRANK = .FALSE.
+!
+      DO I=1, NRGRD
+
+        DO J=1, NRGRD
+          STORES(I,J)%INIT = .FALSE.
+          STORES(I,J)%NTOT = 0
+          STORES(I,J)%NFIN = 0
+          END DO
+        END DO
+!
+      ITAG   = 0
+!
+! -------------------------------------------------------------------- /
+! 1.  Grid point relations and temp data storage
+! 1.a Outer loop over all grids
+!
+!/T      WRITE (MDST,9010)
+!
+      DO I=1, NRGRD
+
+!/T        WRITE (MDST,9011) I, GRANK(I)
+!
+! 1.b Find sub grids with same rank
+!
+        NR = 0
+!
+        DO J=1, NRGRD
+          IF( (GRANK(I).NE.GRANK(J)) .OR. (I.EQ.J) ) CYCLE
+          SHRANK(I,J) = .TRUE.
+          NR = NR + 1
+        END DO
+!
+        DOGRID(I) = NR .GT. 0
+!
+        IF( NR .EQ. 0 ) CYCLE
+!
+        CALL W3SETG( I, MDSE, MDST )
+!
+! Find local root PE and NAPROC for I grid.
+        ICROOT = MDATAS(I)%CROOT
```

Comment:
@ukmo-jianguo-li - I am getting compile errors with `ww3_multi` when compiling as serial program (i.e. with the SHRD flag, rather than MPI DIST).

This subroutine (`WMSMCEQL`) accesses variable `CROOT`, but this variable is only defined when the MPI flag is set:

https://github.com/NOAA-EMC/WW3/blob/7db2f81f144bd882213798a62145c6cd44ce0442/model/ftn/wmmdatmd.ftn#L995

I think this subroutine needs some `!/MPI` switches added in appropriate places.

You can test this with one of the serial multi-grid regression tests, such as `mww3_test_02/PR1`

Reply:
@ukmo-jianguo-li  There are other problems too with SHRD compilations:

This line is not needed and causes compilation errors in SHRD mode - I think it can simply be deleted:
https://github.com/NOAA-EMC/WW3/blob/7db2f81f144bd882213798a62145c6cd44ce0442/model/ftn/wmgridmd.ftn#L4857

This line is using CROOT too and needs to be behind an `MPI` switch:
https://github.com/NOAA-EMC/WW3/blob/7db2f81f144bd882213798a62145c6cd44ce0442/model/ftn/wminitmd.ftn#L2297

Can you fix these and run the `mww3_test_02` regtest (switch PR1) to check that there any on other issues?